source/json.c

```diff
@@ -0,0 +1,306 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+
+#include <aws/common/byte_buf.h>
+#include <aws/common/string.h>
+
+#include <aws/common/json.h>
+
+#include <aws/common/external/cJSON.h>
+
+static struct aws_allocator *s_aws_json_module_allocator = NULL;
+static bool s_aws_json_module_initialized = false;
+
+struct aws_json_value *aws_json_new_string(struct aws_allocator *allocator, const struct aws_byte_cursor cursor) {
+    struct aws_string *tmp = aws_string_new_from_cursor((struct aws_allocator *)allocator, &cursor);
+    void *ret_val = cJSON_CreateString(aws_string_c_str(tmp));
+    aws_string_destroy_secure(tmp);
+    return ret_val;
+}
+
+struct aws_json_value *aws_json_new_number(struct aws_allocator *allocator, double number) {
+    (void)(allocator); // prevent warnings over unused parameter
+    return (void *)(uintptr_t)cJSON_CreateNumber(number);
+}
+
+struct aws_json_value *aws_json_new_array(struct aws_allocator *allocator) {
+    (void)(allocator); // prevent warnings over unused parameter
+    return (void *)(uintptr_t)cJSON_CreateArray();
+}
+
+struct aws_json_value *aws_json_new_boolean(struct aws_allocator *allocator, const bool boolean) {
+    (void)(allocator); // prevent warnings over unused parameter
+    return (void *)(uintptr_t)cJSON_CreateBool(boolean);
+}
+
+struct aws_json_value *aws_json_new_null(struct aws_allocator *allocator) {
+    (void)(allocator); // prevent warnings over unused parameter
+    return (void *)(uintptr_t)cJSON_CreateNull();
+}
+
+struct aws_json_value *aws_json_new_object(struct aws_allocator *allocator) {
+    (void)(allocator); // prevent warnings over unused parameter
+    return (void *)(uintptr_t)cJSON_CreateObject();
+}
+
+int aws_json_value_get_string(const struct aws_json_value *value, struct aws_byte_cursor *output) {
+    struct cJSON *cjson = (struct cJSON *)value;
+    if (cJSON_IsString(cjson)) {
```

Comment:
If you have a function that returns AWS_OP_ERR/AWS_OP_SUCCESS, and you're returning AWS_OP_ERR, then either you need to aws_raise_error() or be certain that the right error has been raised by what you've called and are reacting to.  Since CJSON does not use aws_raise_error() you'll need to do that here.  I would switch things around like this:

```
if (!cJSON_IsString(cjson)) {
   return aws_raise_error(AWS_ERROR_INVALID_ARGUMENT);
}

*output = aws_byte_cursor(...);
return AWS_OP_SUCCESS;
```

Similar for all the other functions with this pattern: raise error if it hasn't been done and prefer early out on errors until you're in the success state.